### ISSUE TITLE
CrossAppDomainCleanUp.CleanupHelper.RegisterCleanup can not concurrent

### ISSUE THREAD

**Issue by takepara** [Antaris/RazorEngine#325]
Error message
System.IndexOutOfRangeException: Index was outside the bounds of the array.
```
Sample stack trace
at System.Collections.Generic.List`1.Add(T item)
at RazorEngine.Compilation.CrossAppDomainCleanUp.CleanupHelper.RegisterCleanupPath(String path)
at RazorEngine.Compilation.ExecutionContextLessThread.ActionConv.Call(Boolean data)
at RazorEngine.Compilation.ExecutionContextLessThread.CallHelperSafeHelper`2.AsAction()
at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
at RazorEngine.Compilation.ExecutionContextLessThread.DefaultCallAction(Action a)
at RazorEngine.Templating.InvalidatingCachingProvider.CacheTemplate(ICompiledTemplate template, ITemplateKey templateKey)
at RazorEngine.Templating.RazorEngineService.CompileAndCacheInternal(ITemplateKey key, Type modelType)
at RazorEngine.Templating.RazorEngineService.GetCompiledTemplate(ITemplateKey key, Type modelType, Boolean compileOnCacheMiss)
at RazorEngine.Templating.RazorEngineService.RunCompile(ITemplateKey key, TextWriter writer, Type modelType, Object model, DynamicViewBag viewBag)
```

I take this stacktrace very very sometimes at Web site project.
CleanupHelper's _toCleanup is not cuncurrent.
Because I changed DefaultCachingProvider constructor default to (_)=>{}, or CacheTemplate in lock block mu wrote custom CachingProvider.
I think..._toCleanup change type to CuncurrentBag.How about?

**Comment by Jaben** [comment]
Having this same issue in production. Big problem.

**Comment by matthid** [comment]
fixed in 3.7.4

**Comment by takepara** [comment]
Thanks matthid!
